src/core/ext/transport/binder/wire_format/wire_writer.cc

```diff
@@ -30,60 +30,160 @@ namespace grpc_binder {
+      return false;
+    }
+  } while (num_outgoing_bytes_ >=
+           num_acknowledged_bytes_ + kFlowControlWindowSize);
+  return true;
+}
+
+int WireWriterImpl::GetFlagForCurrentTransaction(int original_flags,
+                                                 int sent_length,
+                                                 int remaining_length) const {
+  int flags = kFlagMessageData;
+  if (sent_length == 0) {
+    // This is the first transaction. Include initial metadata if there's any.
+    if (original_flags & kFlagPrefix) {
+      flags |= kFlagPrefix;
+    }
+  }
+  if (remaining_length == 0) {
+    // This is the last transaction. Include trailing metadata if there's any.
+    if (original_flags & kFlagSuffix) {
+      flags |= kFlagSuffix;
+    }
+  } else {
+    // There are more messages to send.
+    flags |= kFlagMessageDataIsPartial;
+  }
+  return flags;
+}
+
 absl::Status WireWriterImpl::RpcCall(const Transaction& tx) {
   // TODO(mingcl): check tx_code <= last call id
   grpc_core::MutexLock lock(&mu_);
   GPR_ASSERT(tx.GetTxCode() >= kFirstCallId);
-  RETURN_IF_ERROR(binder_->PrepareTransaction());
-  WritableParcel* parcel = binder_->GetWritableParcel();
-  {
-    //  fill parcel
-    RETURN_IF_ERROR(parcel->WriteInt32(tx.GetFlags()));
-    RETURN_IF_ERROR(parcel->WriteInt32(tx.GetSeqNum()));
-    if (tx.GetFlags() & kFlagPrefix) {
-      // prefix set
-      if (tx.IsClient()) {
-        // Only client sends method ref.
-        RETURN_IF_ERROR(parcel->WriteString(tx.GetMethodRef()));
-      }
-      RETURN_IF_ERROR(parcel->WriteInt32(tx.GetPrefixMetadata().size()));
-      for (const auto& md : tx.GetPrefixMetadata()) {
-        RETURN_IF_ERROR(parcel->WriteByteArrayWithLength(md.first));
-        RETURN_IF_ERROR(parcel->WriteByteArrayWithLength(md.second));
-      }
+  if (CanBeSentInOneTransaction(tx)) {
+    return RpcCallFastPath(tx);
+  }
+  // Slow path: the message data is too large to fit in one transaction.
+  int& seq = seq_num_[tx.GetTxCode()];
+  int original_flags = tx.GetFlags();
+  GPR_ASSERT(original_flags & kFlagMessageData);
+  absl::string_view data = tx.GetMessageData();
+  size_t ptr = 0;
+  while (ptr < data.size()) {
+    if (!WaitForAcknowledgement()) {
+      return absl::InternalError("Timeout waiting for acknowledgement");
     }
-    if (tx.GetFlags() & kFlagMessageData) {
-      RETURN_IF_ERROR(parcel->WriteByteArrayWithLength(tx.GetMessageData()));
+    RETURN_IF_ERROR(binder_->PrepareTransaction());
+    WritableParcel* parcel = binder_->GetWritableParcel();
+    size_t size = std::min(static_cast<size_t>(kBlockSize), data.size() - ptr);
+    int flags = GetFlagForCurrentTransaction(
+        original_flags, /*sent_length=*/ptr,
+        /*remaining_length=*/data.size() - ptr - size);
+    RETURN_IF_ERROR(parcel->WriteInt32(flags));
+    RETURN_IF_ERROR(parcel->WriteInt32(seq++));
+    if (flags & kFlagPrefix) {
+      RETURN_IF_ERROR(WriteInitialMetadata(tx, parcel));
```

Comment:
Would it makes sense to always send initial metadata and trailing metadata separately? i.e Don't put them and message data in the same transaction.

I guess we probably should consider this after the `AParcel_getsize` PR

Reply:
It would work, but I'm not sure why it will be beneficial to send them separately. Seems like the Java implementation also pack them together if possible: https://github.com/grpc/grpc-java/blob/master/binder/src/main/java/io/grpc/binder/internal/Outbound.java#L222.